Vue Integration

CMCDragonkai opened the pull request:
Fixing #1 and #2.

We are 99% there.

We are just missing something about vue template compilation.

Also we would like to have:

1. JSX support - not sure if this is possible with Vue, seems like Vue prefers to use its own syntax style... if so, oh well
2. Just need to optimize some options in webpack.config.js. The output and devserver options have always been a bit confusing for me
3. Eventually get the actual full hello world going.

This deviates from master which @DrFacepalm was using the vue2 CLI for. This is actually using vue3, from first principles.

* Webpack
* Typescript

That's it. No babel.

More things to read here: https://v3.vuejs.org/guide/installation.html#from-cdn-or-without-a-bundler

 To run this.

```
npm run server
```

![image](https://user-images.githubusercontent.com/640797/94016754-e1797300-fdf1-11ea-854a-92c93d876cb6.png)

CMCDragonkai committed:
Add initial SPA demo

CMCDragonkai commented:
For @DrFacepalm to continue.

Also @robert-cronin you should watch this as well, as this will be relevant for Polykey GUI. And also... that one will be even more complex as that's using Electron for Desktop and NativeScript for Mobile.

CMCDragonkai commented:
Key thing to understand about this SPA demo, is that it's a demo of SPA. An SPA app is not npm installable unlike typescript-demo-lib. An SPA app actually needs to be "hosted" by a web server.

The more sophisticated one is `typescript-demo-iso` which will then need to have both frontend and backend done.

Whereas Polykey is none of these, but will use some of the ideas from iso (involving multiple webpack targets) and this one.

CMCDragonkai commented:
Key thing to realize with spa demo. There are NO `dependencies`. Everything is in `devDependencies`. All libraries are bundled. Although the bundler may choose to alias certain libraries to CDN links.

DrFacepalm commented:
Just pushed up changes to `webpack.config.json`, i've added an exports default app in index.js (tryign to get it to work with testing). The main change here is that:
* in the `appConfig`, in `resolve`, I've added: `alias: { 'vue': 'vue/dist/vue.esm-bundler.js' }`
* in plugins: ive added: 
```js
new webpack.DefinePlugin({
      __VUE_OPTIONS_API__: true, // enable/disable Options API support,
      __VUE_PROD_DEVTOOLS__: false // enable/disable devtools support in production
})
```
sourced from: https://github.com/vuejs/vue-next/tree/master/packages/vue#bundler-build-feature-flags

CMCDragonkai commented:
No extra dependencies needed?

On 9/24/20 8:52 PM, DrFacepalm wrote:
>
> Just pushed up changes to |webpack.config.json|, i've added an exports 
> default app in index.js (tryign to get it to work with testing). The 
> main change here is that:
>
>   * in the |appConfig|, in |resolve|, I've added: |alias: { 'vue':
>     'vue/dist/vue.esm-bundler.js' }|
>   * in plugins: ive added:
>
> new  webpack.DefinePlugin({
>        __VUE_OPTIONS_API__:true,  // enable/disable Options API support,
>        __VUE_PROD_DEVTOOLS__:false  // enable/disable devtools support in production
> })
>
> sourced from: 
> https://github.com/vuejs/vue-next/tree/master/packages/vue#bundler-build-feature-flags
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MatrixAI/Typescript-Demo-Spa/pull/3#issuecomment-698269421>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAE4OHIPPTXAEX2HPM2NPHDSHMQGHANCNFSM4RXASLHQ>.
>

DrFacepalm commented:
> No extra dependencies needed?
> […](#)
> On 9/24/20 8:52 PM, DrFacepalm wrote: Just pushed up changes to |webpack.config.json|, i've added an exports default app in index.js (tryign to get it to work with testing). The main change here is that: * in the |appConfig|, in |resolve|, I've added: |alias: { 'vue': 'vue/dist/vue.esm-bundler.js' }| * in plugins: ive added: new webpack.DefinePlugin({ __VUE_OPTIONS_API__:true, // enable/disable Options API support, __VUE_PROD_DEVTOOLS__:false // enable/disable devtools support in production }) sourced from: https://github.com/vuejs/vue-next/tree/master/packages/vue#bundler-build-feature-flags — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[#3 (comment)](https://github.com/MatrixAI/Typescript-Demo-Spa/pull/3#issuecomment-698269421)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAE4OHIPPTXAEX2HPM2NPHDSHMQGHANCNFSM4RXASLHQ>.

No, webpack's DefinePlugin is built in so i just had to bring in webpack. which we already have. Didn't have to do any extra installs

CMCDragonkai commented:
So that enables dynamic compilation? What if we want webpack to compile the vue templates ahead of time assuming we aren't using JSX. But I'm still interested in seeing if JSX can still be used.

DrFacepalm commented:
I think the .vue files are done such that their render() function is compiled at build time, so if that's what you're talking about then there's an extra webpack dependency: https://vue-loader.vuejs.org/guide/#vue-cli. I'll have to make and play around with some .vue files to se how it works.

CMCDragonkai commented:
So vue SFC is now working with the latest `vue-loader`. However the `index.ts` had to be changed to `index.js`. So please investigate how to use `index.ts` instead. Or to make our `ts-loader` work properly with `vue-loader`.

So it seems like something needs to be done to the `vue-loader` in order for it make understanding of the typescript files.

CMCDragonkai commented:
Following:

* https://dev.to/lmillucci/building-a-vue-3-component-with-typescript-4pge
* https://blog.logrocket.com/new-features-in-vue-3-and-how-to-use-them/

The `index.ts` now works.

@DrFacepalm needed the sfc to compile the SFC.

The last thing is to check how to bring in CSS into the vue.

At this point you may need the CSS loader. But try it anyway.

CMCDragonkai commented:
![image](https://user-images.githubusercontent.com/640797/94222532-fad50900-ff30-11ea-8443-f3d6c8518717.png)

CMCDragonkai commented:
Now we can use `ts` inside the SFC, and that required the:

```
        options: {
          appendTsSuffixTo: [/\.vue$/]
        }
```

So @DrFacepalm needs to test 2 things:

1. Bringing in CSS
2. Bringing in images

These may require extra loaders. Not sure. Just test them.

It looks like we are not using JSX at all. That's fine. It seems vue is meant to be using these kinds of SFCs.

CMCDragonkai commented:
Note that it's important to realize that https://www.npmjs.com/package/vue-loader is needed, but atm it requires a `16.0.0` version which then supports the `"@vue/compiler-sfc": "^3.0.0"`.

Atm, it's not yet released. It's still under the next alias. But we are pinning it in the `package.json`. When `16.0.0` is properly released, then we can actually just use it normally.

DrFacepalm commented:
Using this repo: https://github.com/vuejs/vue-next-webpack-preview as a base, there is now a css loader.
We're using `MiniCssExtractPlugin` and we've also had to install `css-loader`

DrFacepalm commented:
https://webpack.js.org/guides/asset-management/#loading-images

webpack has a file-loader so we just have to `npm install -D file-loader` and then add a rule to use `file-loader` for `(png|svg|jpg|gif)` files.

DrFacepalm commented:
![image](https://user-images.githubusercontent.com/31267481/94224579-23abcd00-ff36-11ea-9729-519aad2e50a3.png)

CMCDragonkai commented:
So this should be ready for merge soon?

Only thing is:

* Testing
* Storybook

DrFacepalm commented:
For testing, there's a framework `@vue/test-utils`: https://vue-test-utils.vuejs.org/guides/#getting-started. Going to bring that in to try out.

DrFacepalm commented:
I've changed the jest config:
Added:
```js
  "transform": {
    "^.+\\.vue$": "vue-jest",
   ...
  },
  moduleFileExtensions: ['vue', 'js', 'json', 'jsx', 'ts', 'tsx', 'node'],
```

Also added to package.json:
* `"@vue/test-utils": "^2.0.0-beta.5"`
* `"vue-jest": "^5.0.0-alpha.4"`

There are some problems however. The `index.test.js` imports App, and tries to mount it using `@vue/test-utils`'s `mount` but it throws an error when it encounters the image. Removing the image from the component's template runs the test, but still has a warning in index.ts saying "Failed to mount app: mount target selector returned null".

DrFacepalm commented:
I've looked into storybook but for vue3, it seems like the support is still being developed. 
* https://github.com/storybookjs/storybook/issues/10654

This issue goes into the progress of that.

There is a work in progress package released here: https://www.npmjs.com/package/@andoshin11/storybook-vue3 though.

DrFacepalm commented:
This should be ready to merge.

CMCDragonkai commented:
Tests should still be written in typescript. Are you sure the test are written properly?

CMCDragonkai commented:
Can you ignore the `.vscode` stuff.

CMCDragonkai commented on jest.config.js:
```diff
@@ -10,8 +10,10 @@ module.exports = {
   ],
   "transform": {
+    "^.+\\.vue$": "vue-jest",
     "^.+\\.tsx?$": "ts-jest"
```
We are no longer using `tsx` at all. Cause there's no JSX support at all.

DrFacepalm replied:
I tried removing it but trying to run the test without it gives: 
![image](https://user-images.githubusercontent.com/31267481/94436644-08fc8100-01e0-11eb-8471-89877e8f03df.png)

CMCDragonkai commented on jest.config.js:
```diff
@@ -10,8 +10,10 @@ module.exports = {
+    "^.+\\.vue$": "vue-jest",
     "^.+\\.tsx?$": "ts-jest"
   },
+  moduleFileExtensions: ['vue', 'js', 'json', 'jsx', 'ts', 'tsx', 'node'],
```
No more `jsx` nor `tsx` and what's `node` for?

DrFacepalm replied:
Removed the unused extensions

CMCDragonkai commented on tsconfig.json:
```diff
@@ -2,18 +2,17 @@
-    "module": "CommonJS",
-    "target": "ES2016",
+    "module": "esnext",
+    "target": "esnext",
```
The target should be targeting something that most browsers can still use. I think ES2016 is fine.

DrFacepalm replied:
Changed target back to ES2016

CMCDragonkai commented on tsconfig.json:
```diff
@@ -2,18 +2,17 @@
     "moduleResolution": "node",
-    "module": "CommonJS",
-    "target": "ES2016",
+    "module": "esnext",
```
Targeting ES6 is probably fine. Commonjs is necessary for typescript-demo-lib because they need to produce libraries for node-consumption. @robert-cronin see this.

DrFacepalm replied:
Changed the module back to ComonJS.

CMCDragonkai commented on webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+    publicPath: '/'
+  },
+  resolve: {
+    extensions: ['.tsx', '.ts', '.js', '.vue', '.json'],
```
No more `tsx`. No longer using JSX atm.

DrFacepalm replied:
Removed unused extensions. now just `ts, js, vue, json`

CMCDragonkai commented on webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+        loader: 'vue-loader'
+      },
+      {
+        test: /.tsx?$/,
```
Same here...

DrFacepalm replied:
Changed to `.ts`

CMCDragonkai commented on webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+        use: [ MiniCssExtractPlugin.loader, 'css-loader' ]
+      },
+      {
+        test: /\.(png|svg|jpg|gif)$/,
```
Do these cover all the image types that we use in the web? Probably not... but it's fine for now.

But what about font files?

DrFacepalm replied:
I just did some quick reading, it seems like local font files should work with `file-loader`, thought it would need some more config: 
https://chriscourses.com/blog/loading-fonts-webpack

Shall i add that in too?

CMCDragonkai replied:
You did already.

CMCDragonkai commented:
No commonjs is only for library. For spa es6 is fine.

On 28 September 2020 11:22:47 pm AEST, DrFacepalm <notifications@github.com> wrote:
>@DrFacepalm commented on this pull request.
>
>
>
>>      "esModuleInterop": true,
>     "allowSyntheticDefaultImports": true,
>     "moduleResolution": "node",
>-    "module": "CommonJS",
>-    "target": "ES2016",
>+    "module": "esnext",
>
>Changed the module back to ComonJS.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/MatrixAI/Typescript-Demo-Spa/pull/3#discussion_r495934518

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

CMCDragonkai commented:
Yea.

On 28 September 2020 11:28:04 pm AEST, DrFacepalm <notifications@github.com> wrote:
>@DrFacepalm commented on this pull request.
>
>
>
>> +        test: /\.vue$/,
>+        loader: 'vue-loader'
>+      },
>+      {
>+        test: /.tsx?$/,
>+        loader: 'ts-loader',
>+        options: {
>+          appendTsSuffixTo: [/\.vue$/]
>+        }
>+      },
>+      {
>+        test: /\.css$/i,
>+        use: [ MiniCssExtractPlugin.loader, 'css-loader' ]
>+      },
>+      {
>+        test: /\.(png|svg|jpg|gif)$/,
>
>I just did some quick reading, it seems like local font files should
>work with `file-loader`, thought it would need some more config: 
>https://chriscourses.com/blog/loading-fonts-webpack
>
>Shall i add that in too?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/MatrixAI/Typescript-Demo-Spa/pull/3#discussion_r495938220

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

DrFacepalm commented:
in tsconfig.json:
```
    "module": "es6",
    "target": "ES2016",
```

in webpack.config.js:
```
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/,
        use: [
          'file-loader'
        ]
      },
```
This handles font files.
